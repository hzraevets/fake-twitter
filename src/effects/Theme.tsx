import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useCallback,
  createContext,
  useState,
  ReactNode,
} from 'react';
import { ConfigProvider, theme as antdTheme } from "antd";

interface ThemeContextInterface {
  theme: string;
  setTheme: Dispatch<SetStateAction<string>>;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextInterface>({
  theme: 'light',
  setTheme: () => null,
  toggleTheme: () => null,
});

const getTheme = () => {
  const theme = localStorage.getItem('theme');
  if (!theme) { // default theme
    localStorage.setItem('theme', 'light');
    return 'light';
  }

  return theme;
};

const ThemeProvider = ({ children }: { children: ReactNode }) => {
  const [ theme, setTheme ] = useState(getTheme);

  function toggleTheme() {
    if (theme === 'dark') {
      setTheme('light');
    } else {
      setTheme('dark');
    }
  }

  const refreshTheme = useCallback(() => {
    localStorage.setItem('theme', theme);
  }, [theme]);

  useEffect(() => {
    refreshTheme();
  }, [refreshTheme, theme]);

  return (
    <ThemeContext.Provider
      value={{
        theme,
        setTheme,
        toggleTheme,
      }}
    >
      <div className={`theme-container ${theme}`}>
        <ConfigProvider
          theme={{
            algorithm: theme === 'dark'
              ? antdTheme.darkAlgorithm
              : antdTheme.defaultAlgorithm
          }}
        >
          {children}
        </ConfigProvider>
      </div>
    </ThemeContext.Provider>
  );
};

export { ThemeContext, ThemeProvider };
