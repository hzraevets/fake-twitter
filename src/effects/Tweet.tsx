import React, { createContext, useState, ReactNode } from 'react';
import getUnixTime from 'date-fns/getUnixTime';

import {
  Tweet,
  TweetStorage,
  CreateNewTweetParams,
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
} from 'models';

export interface TweetContextInterface {
  allTweets: TweetStorage;
  nextTweetId: number;
  addNewTweet: (params: CreateNewTweetParams) => void;
  readTimeline: (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>;
  editTweet: (id: number, content: string) => void;
  deleteTweet: (id: number) => void;
}

const initialTweetStorage: TweetStorage = {
  hashMap: {},
  order: [],
};

const TweetContext = createContext<TweetContextInterface>({
  allTweets: initialTweetStorage,
  nextTweetId: 1,
  addNewTweet: () => null,
  readTimeline: () => Promise.resolve({ data: [] }),
  editTweet: () => null,
  deleteTweet: () => null,
});

const TweetProvider = ({ children }: { children: ReactNode }) => {
  const [ allTweets, setAllTweets ] = useState(initialTweetStorage);
  const [ nextTweetId, setNextTweetId ] = useState(1);

  function addNewTweet(params: CreateNewTweetParams) {
    const now = getUnixTime(new Date());
    const { hashMap, order } = allTweets;

    const newTweet: Tweet = {
      ...params,
      id: nextTweetId,
      createTime: now,
      updateTime: now,
      isDeleted: false,
    };

    setNextTweetId(nextTweetId + 1);
    setAllTweets({
      hashMap: { ...hashMap, [nextTweetId]: newTweet },
      order: [ nextTweetId, ...order ],
    });
  }

  function readTimeline({ start, offset = 1 }: ReadNonDeleteTweetParams) {
    const data: Tweet[] = [];
    let nextCursor: number | null;

    const { hashMap, order } = allTweets;

    for (nextCursor = start; (data.length < offset) && nextCursor < order.length; nextCursor++) {
      const tweet = hashMap[order[nextCursor]];

      if (!tweet.isDeleted) {
        data.push(tweet);
      }
    }

    const res: ReadNonDeleteTweetResponse = { data };

    if (nextCursor < order.length) {
      res.nextCursor = nextCursor;
    }

    return new Promise<ReadNonDeleteTweetResponse>(resolve =>
      setTimeout(() => resolve(res), Math.random() * 1000));
  }

  function editTweet(id: number, content: string) {
    const { hashMap, order } = allTweets;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].content = content;
    hashMap[id].updateTime = getUnixTime(new Date());

    setAllTweets({ hashMap, order });
  }

  function deleteTweet(id: number) {
    const { hashMap } = allTweets;
    let order = allTweets.order;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].isDeleted = true;
    order = order.filter(tid => tid !== id);

    setAllTweets({ hashMap, order });
  }

  return (
    <TweetContext.Provider
      value={{
        allTweets,
        nextTweetId,
        addNewTweet,
        readTimeline,
        editTweet,
        deleteTweet,
      }}
    >
      {children}
    </TweetContext.Provider>
  );
}

export { TweetContext, TweetProvider };
