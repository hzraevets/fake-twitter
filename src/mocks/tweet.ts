import {
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
  Tweet,
  TweetStorage,
} from 'models';

export const testUserTweet1: Tweet = {
  id: 1,
  author: 'testUser',
  firstname: 'Test',
  content: 'Hello, world!',
  createTime: 1677073172,
  updateTime: 1677073172,
  isDeleted: false,
};

export const liBaiTweet1: Tweet = {
  id: 2,
  author: 'liBai',
  firstname: 'Li',
  content: '抽刀断水水更流，举杯销愁愁更愁',
  createTime: 1677250735,
  updateTime: 1677250775,
  isDeleted: false,
};

export const storage: TweetStorage = {
  hashMap: {
    [testUserTweet1.id]: testUserTweet1,
    [liBaiTweet1.id]: liBaiTweet1,
  },
  order: [
    liBaiTweet1.id,
    testUserTweet1.id,
  ],
};

export const nextTweetId = 3;

export const mockedReadTimeline = ({ start, offset = 1 }: ReadNonDeleteTweetParams) => {
  const dat: Tweet[] = [];
  let nextCursor: number | null;

  const { hashMap, order } = storage;

  for (nextCursor = start; (dat.length < offset) && nextCursor < order.length; nextCursor++) {
    const tweet = hashMap[order[nextCursor]];

    if (!tweet.isDeleted) {
      dat.push(tweet);
    }
  }

  const res: ReadNonDeleteTweetResponse = { data: dat };

  if (nextCursor < order.length) {
    res.nextCursor = nextCursor;
  }

  return Promise.resolve<ReadNonDeleteTweetResponse>(res);
}

