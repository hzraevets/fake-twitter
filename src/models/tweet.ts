export interface Tweet {
  id: number;
  author: string;
  firstname: string;
  content: string;
  createTime: number; // unix timestamp
  updateTime: number; // unix timestamp
  isDeleted: boolean;
}

export interface TweetStorage {
  hashMap: { [id: number]: Tweet };
  order: number[]; // desc for tweet according to create time
}

export type CreateNewTweetParams = Omit<
  Tweet,
  'id' | 'createTime' | 'updateTime' | 'isDeleted'
>;

export interface ReadNonDeleteTweetParams {
  start: number;
  offset?: number;
}

export interface ReadNonDeleteTweetResponse {
  data: Tweet[];
  nextCursor?: number | null;
}

export type UpdateTweetParams = Omit<
  Tweet,
  'firstname' | 'createTime' | 'updateTime' | 'isDeleted'
>;

export type DeleteTweetParams = Omit<
  Tweet,
  'content' | 'firstname' | 'createTime' | 'updateTime' | 'isDeleted'
>;
