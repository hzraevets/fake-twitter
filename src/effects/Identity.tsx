import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  createContext,
  useState,
  ReactNode,
} from 'react';

import { UserStorage, User } from 'models';
import { md5 } from 'utils/md5';

export interface IdentityContextInterface {
  identity: UserStorage;
  setIdentity: Dispatch<SetStateAction<UserStorage>>;
  loginUser: string | null;
  setLoginUser: Dispatch<SetStateAction<string | null>>;
  logout: () => void;
  register: (newUser: User) => UserStorage;
}

const IdentityContext = createContext<IdentityContextInterface>({
  identity: {},
  setIdentity: () => null,
  loginUser: null,
  setLoginUser: () => null,
  logout: () => null,
  register: () => ({}),
});

const getIdentity = (): UserStorage => {
  const identity = localStorage.getItem('identity');
  if (!identity) {
    localStorage.setItem('identity', '{}');
    return {};
  }

  return JSON.parse(identity);
};

const getLoginUser = (): string | null => {
  const loginUser = localStorage.getItem('loginUser');
  if (!loginUser) {
    return null;
  }

  return loginUser;
};

const IdentityProvider = ({ children }: { children: ReactNode }) => {
  const [ identity, setIdentity ] = useState(getIdentity);
  const [ loginUser, setLoginUser ] = useState(getLoginUser);

  function logout() {
    if (loginUser) {
      setLoginUser(null);
      localStorage.removeItem('loginUser');
    }
  }

  function register(newUser: User) {
    const newStorage: UserStorage = {
      ...identity,
      [newUser.username]: {
        ...newUser,
        password: md5(newUser.password),
      },
    };

    localStorage.setItem('identity', JSON.stringify(newStorage));
    return newStorage;
  }

  useEffect(() => {
    const updateIdentity = () => {
      localStorage.setItem('identity', JSON.stringify(identity));
    };

    updateIdentity();
  }, [identity]);

  useEffect(() => {
    const executeLogin = () => {
      if (loginUser) {
        localStorage.setItem('loginUser', loginUser);
      }
    };

    executeLogin();
  }, [loginUser]);

  return (
    <IdentityContext.Provider
      value={{
        identity,
        setIdentity,
        loginUser,
        setLoginUser,
        logout,
        register,
      }}
    >
      {children}
    </IdentityContext.Provider>
  );
};

export { IdentityContext, IdentityProvider };
