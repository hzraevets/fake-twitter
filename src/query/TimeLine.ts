import { QueryClient } from 'react-query';

import { ReadNonDeleteTweetParams, ReadNonDeleteTweetResponse } from 'models';

export const timeLineQuery = new QueryClient();

export const generateFetchTimeline = (
  readTimeline:  (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>
) => ({ pageParam = 0 }) => {
  return readTimeline({ start: pageParam });
}

export const clearThenReFetch = async (
  readTimeline:  (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>
) => {
  timeLineQuery.clear();

  await setTimeout(() => null, 500);

  return timeLineQuery.prefetchInfiniteQuery('timeline', generateFetchTimeline(readTimeline), {
    getNextPageParam: (lastPage, pages) => {
      return lastPage.nextCursor;
    },
  });
}
