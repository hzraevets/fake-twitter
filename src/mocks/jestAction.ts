import { enableLocalStorageMock, enableMatchMedia } from 'utils/browserMocks';
import { TweetContextInterface } from 'effects/Tweet';
import { storage as identityStorage, testUser } from './identity';
import { nextTweetId, storage as tweetStorage, mockedReadTimeline } from './tweet';

export function prepareTweetProviderPropsWithoutCRUD(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: () => Promise.resolve({ data: [] }),
    editTweet: () => null,
    deleteTweet: () => null,
  };
}

export function prepareTweetProviderPropsWithRead(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: mockedReadTimeline,
    editTweet: () => null,
    deleteTweet: () => null,
  };
}
