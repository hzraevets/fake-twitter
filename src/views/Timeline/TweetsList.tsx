import React, { useContext, useRef, useEffect, useMemo } from 'react';
import { useInfiniteQuery } from 'react-query';
import { Card, Spin, Skeleton, Divider, List, Avatar } from 'antd';
import InfiniteScroll from 'react-infinite-scroll-component';
import { UserOutlined } from '@ant-design/icons';
import formatDistance from 'date-fns/formatDistance';
import fromUnixTime from 'date-fns/fromUnixTime';
import { Link } from 'react-router-dom';

import { TweetContext } from 'effects/Tweet';
import { ReadNonDeleteTweetResponse } from 'models';
import {
  generateFetchTimeline,
  getNextPageParam,
  flattenQueryPages,
} from 'query/TimeLine';
import './tweets-list.css';

export function TweetsList() {
  const elementRef = useRef<HTMLDivElement>(null);
  const { readTimeline } = useContext(TweetContext);
  const fetchTimeline = generateFetchTimeline(readTimeline);

  const {
    data,
    fetchNextPage,
    hasNextPage,
    status,
  } = useInfiniteQuery<ReadNonDeleteTweetResponse>('timeline', fetchTimeline, {
    getNextPageParam,
  });

  useEffect(() => {
    // https://stackoverflow.com/questions/7668636/check-with-jquery-if-div-has-overflowing-elements/7668692#7668692
    const cardElement = elementRef.current;

    const offsetTop = cardElement?.offsetTop || 0;
    const offsetHeight = cardElement?.offsetHeight || 0;
    const windowHeight = window.innerHeight;

    if (hasNextPage && (offsetTop + offsetHeight < windowHeight)) {
      fetchNextPage();
    }
  });

  const now = new Date();

  const flattenTweetList = useMemo(
    () => flattenQueryPages(data?.pages || []),
    [data?.pages],
  );

  return (
    <Card className="w-80 overflow-auto break-inside-avoid-column mb-10" ref={elementRef}>
      {status === 'loading' && <div className="w-full flex"><Spin className="m-auto"/></div>}
      {status === 'error' && <p>Unknown Error</p>}
      {status !== 'loading' && status !== 'error' && (
        <InfiniteScroll
          dataLength={flattenTweetList.length}
          next={fetchNextPage}
          hasMore={hasNextPage || false}
          loader={<Skeleton avatar paragraph={{ rows: 1 }} active />}
          endMessage={<Divider plain className="select-none">It is all, nothing more ü§ê</Divider>}
          scrollableTarget="scrollableDiv"
        >
          {flattenTweetList.length > 0 ? (
            <List
              dataSource={flattenTweetList}
              renderItem={(tweet, index) => (
                <List.Item key={`${index}-${tweet.id}`}>
                  <Link to={`/tweet/${tweet.id}`} className="w-full">
                    <List.Item.Meta
                      avatar={<Avatar size="large" className="bg-slate-500" icon={<UserOutlined/>} />}
                      title={(
                        <div className="select-none">
                          <span className="first-name text-sm mr-1">{tweet.firstname}</span>
                          {(tweet.updateTime !== tweet.createTime) && (
                            <span className="text-xs text-slate-400">edited</span>
                          )}
                          <span className="text-xs text-slate-500 float-right">
                              {formatDistance(fromUnixTime(tweet.createTime), now, { addSuffix: true })}
                            </span>
                        </div>
                      )}
                      description={(
                        <p className="m-0 break-words text-sm">{tweet.content}</p>
                      )}
                    />
                  </Link>
                </List.Item>
              )}
            />
          ) : null}
        </InfiniteScroll>
      )}
    </Card>
  );
}
