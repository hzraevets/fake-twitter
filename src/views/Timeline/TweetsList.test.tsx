import React from 'react';
import { act } from 'react-dom/test-utils';
import {render, screen} from '@testing-library/react';
import { QueryClientProvider } from 'react-query';
// @ts-ignore
import { create } from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { TweetsList } from './TweetsList';
import { timeLineQuery } from 'query/TimeLine';
import { TweetContext } from 'effects/Tweet';
import { prepareTweetProviderPropsWithRead } from 'mocks/jestAction';
import { testUserTweet1, liBaiTweet1 } from 'mocks/tweet';

describe('TweetsList', () => {
  it('render empty TweetList component', async () => {
    /**
     * When testing, code that causes React state updates should be wrapped into act(...):
     *
     *     act(() => {
     *       // fire events that update state
     *     });
     *     // assert on the output
     *
     * This ensures that you're testing the behavior the user would see in the browser.
     * Learn more at https://reactjs.org/link/wrap-tests-with-act
     *
     * also see https://github.com/react-hook-form/react-hook-form/discussions/7125
     **/
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      render(
        <BrowserRouter>
          <QueryClientProvider client={timeLineQuery}>
            <TweetsList/>
          </QueryClientProvider>
        </BrowserRouter>
      );
    });
  });

  it('render non-empty TweetList component', async () => {
    const providerProps = prepareTweetProviderPropsWithRead();

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      render(
        <BrowserRouter>
          <TweetContext.Provider value={providerProps}>
            <QueryClientProvider client={timeLineQuery}>
              <TweetsList/>
            </QueryClientProvider>
          </TweetContext.Provider>
        </BrowserRouter>
      );
    });

    expect(screen.getByText(liBaiTweet1.content)).toBeInTheDocument();
    expect(screen.getByText(testUserTweet1.content)).toBeInTheDocument();
  });

  it('render empty TweetList to snapshot', () => {
    const renderer = create(
      <BrowserRouter>
        <QueryClientProvider client={timeLineQuery}>
          <TweetsList/>
        </QueryClientProvider>
      </BrowserRouter>
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('render non-empty TweetList to snapshot', async () => {
    const providerProps = prepareTweetProviderPropsWithRead();

    const renderer = await act(async () => {
      return create(
        <BrowserRouter>
          <TweetContext.Provider value={providerProps}>
            <QueryClientProvider client={timeLineQuery}>
              <TweetsList/>
            </QueryClientProvider>
          </TweetContext.Provider>
        </BrowserRouter>
      );
    });
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
