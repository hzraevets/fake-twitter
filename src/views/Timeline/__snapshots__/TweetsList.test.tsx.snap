// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TweetsList render empty TweetList to snapshot 1`] = `
<div
  className="ant-card ant-card-bordered w-80 overflow-auto break-inside-avoid-column mb-10 css-dev-only-do-not-override-diro6f"
>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component "
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="ant-list ant-list-split css-dev-only-do-not-override-diro6f"
        >
          <div
            className="ant-spin-nested-loading css-dev-only-do-not-override-diro6f"
          >
            <div
              className="ant-spin-container"
            >
              <ul
                className="ant-list-items"
              >
                <li
                  className="ant-list-item"
                >
                  <a
                    className="w-full"
                    href="/tweet/2"
                    onClick={[Function]}
                  >
                    <div
                      className="ant-list-item-meta"
                    >
                      <div
                        className="ant-list-item-meta-avatar"
                      >
                        <span
                          className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-icon bg-slate-500 css-dev-only-do-not-override-diro6f"
                          style={Object {}}
                        >
                          <span
                            aria-label="user"
                            className="anticon anticon-user"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="user"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                              />
                            </svg>
                          </span>
                        </span>
                      </div>
                      <div
                        className="ant-list-item-meta-content"
                      >
                        <h4
                          className="ant-list-item-meta-title"
                        >
                          <div
                            className="select-none"
                          >
                            <span
                              className="first-name text-sm mr-1"
                            >
                              Li
                            </span>
                            <span
                              className="text-xs text-slate-400"
                            >
                              edited
                            </span>
                            <span
                              className="text-xs text-slate-500 float-right"
                            >
                              2 days ago
                            </span>
                          </div>
                        </h4>
                        <div
                          className="ant-list-item-meta-description"
                        >
                          <p
                            className="m-0 break-words text-sm"
                          >
                            ÊäΩÂàÄÊñ≠Ê∞¥Ê∞¥Êõ¥ÊµÅÔºå‰∏æÊùØÈîÄÊÑÅÊÑÅÊõ¥ÊÑÅ
                          </p>
                        </div>
                      </div>
                    </div>
                  </a>
                </li>
                <li
                  className="ant-list-item"
                >
                  <a
                    className="w-full"
                    href="/tweet/1"
                    onClick={[Function]}
                  >
                    <div
                      className="ant-list-item-meta"
                    >
                      <div
                        className="ant-list-item-meta-avatar"
                      >
                        <span
                          className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-icon bg-slate-500 css-dev-only-do-not-override-diro6f"
                          style={Object {}}
                        >
                          <span
                            aria-label="user"
                            className="anticon anticon-user"
                            role="img"
                          >
                            <svg
                              aria-hidden="true"
                              data-icon="user"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                              />
                            </svg>
                          </span>
                        </span>
                      </div>
                      <div
                        className="ant-list-item-meta-content"
                      >
                        <h4
                          className="ant-list-item-meta-title"
                        >
                          <div
                            className="select-none"
                          >
                            <span
                              className="first-name text-sm mr-1"
                            >
                              Test
                            </span>
                            <span
                              className="text-xs text-slate-500 float-right"
                            >
                              4 days ago
                            </span>
                          </div>
                        </h4>
                        <div
                          className="ant-list-item-meta-description"
                        >
                          <p
                            className="m-0 break-words text-sm"
                          >
                            Hello, world!
                          </p>
                        </div>
                      </div>
                    </div>
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          className="ant-divider css-dev-only-do-not-override-diro6f ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center ant-divider-plain select-none"
          role="separator"
        >
          <span
            className="ant-divider-inner-text"
            style={Object {}}
          >
            It is all, nothing more ü§ê
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TweetsList render non-empty TweetList to snapshot 1`] = `
<div
  className="ant-card ant-card-bordered w-80 overflow-auto break-inside-avoid-column mb-10 css-dev-only-do-not-override-diro6f"
>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component "
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="ant-divider css-dev-only-do-not-override-diro6f ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center ant-divider-plain select-none"
          role="separator"
        >
          <span
            className="ant-divider-inner-text"
            style={Object {}}
          >
            It is all, nothing more ü§ê
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
